rectools/models/sasrec.py

```diff
@@ -556,40 +503,127 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
-    """TODO"""
+class SessionEncoderLightningModuleBase(LightningModule):
+    """Base class for lightning module. Used only for type hinting."""
 
     def __init__(
         self,
-        session_maxlen: int,
+        torch_model: TransformerBasedSessionEncoder,
         lr: float,
-        batch_size: int,
-        epochs: int,
-        device: str,
-        n_blocks: int,
-        n_factors: int,
-        n_heads: int,
-        dropout_rate: float,
-        use_pos_emb: bool = True,
         loss: str = "softmax",
+        adam_betas: Tuple[float, float] = (0.9, 0.98),
+    ):
+        super().__init__()
+        self.lr = lr
+        self.loss = loss
+        self.torch_model = torch_model
+        self.adam_betas = adam_betas
+
+    def configure_optimizers(self) -> torch.optim.Adam:
+        """TODO"""
+        optimizer = torch.optim.Adam(self.torch_model.parameters(), lr=self.lr, betas=self.adam_betas)
+        return optimizer
+
+    def forward(
+        self,
+        batch: torch.Tensor,
+    ) -> torch.Tensor:
+        """TODO"""
+        return self.torch_model(batch)
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        raise NotImplementedError()
+
+
+class SessionEncoderLightningModule(SessionEncoderLightningModuleBase):
+    """TODO"""
+
+    def on_train_start(self) -> None:
+        """TODO"""
+        self._xavier_normal_init()
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        x, y, w = batch
+        logits = self.forward(x)  # [batch_size, session_maxlen, n_items + 1]
+        if self.loss == "softmax":
+            # We are using CrossEntropyLoss with a multi-dimensional case
+
+            # Logits must be passed in form of [batch_size, n_items + 1, session_maxlen],
+            #  where n_items + 1 is number of classes
+
+            # Target label indexes must be passed in a form of [batch_size, session_maxlen]
+            # (`0` index for "PAD" ix excluded from loss)
+
+            # Loss output will have a shape of [batch_size, session_maxlen]
+            # and will have zeros for every `0` target label
+
+            loss = torch.nn.functional.cross_entropy(
+                logits.transpose(1, 2), y, ignore_index=0, reduction="none"
+            )  # [batch_size, session_maxlen]
+            loss = loss * w
+            n = (loss > 0).to(loss.dtype)
+            loss = torch.sum(loss) / torch.sum(n)
+            return loss
+        raise ValueError(f"loss {loss} is not supported")
+
+    def _xavier_normal_init(self) -> None:
+        """TODO"""
+        for _, param in self.torch_model.named_parameters():
+            try:
+                torch.nn.init.xavier_normal_(param.data)
+            except ValueError:
+                pass
+
+
+# ####  --------------  SASRec Model  --------------  #### #
+
+
+class SasRecModel(ModelBase):
```

Comment:
```suggestion
class SASRecModel(ModelBase):
```
Let's change it now until it's too late :)

Reply:
fixed